app.go

```diff
@@ -453,28 +464,92 @@ func ValidateApp(opts ...Option) error {
 // registered via Invoke options. See the documentation of the App struct for
 // details on the application's initialization, startup, and shutdown logic.
 func New(opts ...Option) *App {
-	logger := fxlog.DefaultLogger(os.Stderr)
-
 	app := &App{
-		log:          logger,
+		log:          new(logBuffer),
 		startTimeout: DefaultTimeout,
 		stopTimeout:  DefaultTimeout,
 	}
 
 	for _, opt := range opts {
 		opt.apply(app)
 	}
-	app.lifecycle = &lifecycleWrapper{lifecycle.New(app.log)}
 
 	app.container = dig.New(
 		dig.DeferAcyclicVerification(),
 		dig.DryRun(app.validate),
 	)
+	var connectLogger func(logger fxlog.Logger) error
+	if app.fxlogConstructor != nil {
+		// Since user supplied a custom logger, use a buffered logger to hold
+		// all messages until user supplied logger is instantiated. Then we flush
+		// those messages after fully constructing the custom logger.
+		var buffer logBuffer
+		app.log = &buffer
+		connectLogger = buffer.Connect // func(fxlog.Logger) error
+	} else {
+		// If user hasn't supplied a custom implementation of fxlog.Logger
+		// then use default one.
+		app.log = fxlog.DefaultLogger(os.Stderr)
+	}
 
 	for _, p := range app.provides {
 		app.provide(p)
 	}
 
+	if app.err != nil {
+		return app
+	}
+
+	// If WithLogger and Printer are both provided, WithLogger takes precedence.
+	// If user supplied an fxlog constructor, we need to have dig's container
+	// initialize it and grab all of arguments from the graph and then return.
+	if app.fxlogConstructor != nil {
+		// Here we add the custom logger into the container.
+		if err := app.container.Provide(app.fxlogConstructor); err != nil {
+			app.err = multierr.Combine(app.err,
+				fmt.Errorf("could not construct customer logger via fx.WithLogger: %w", err))
+			if connectLogger != nil {
+				err = connectLogger(fxlog.DefaultLogger(os.Stderr))
+				app.err = multierr.Combine(app.err, err)
+			} else {
+				// This should not happen as connectLogger should have been defined in
+				// if app.fxlogConstructor != nil  branch above.
+				err = fmt.Errorf("internal error: connectLogger should have been initialized")
+				app.err = multierr.Combine(app.err, err)
+			}
+
+			return app
+		}
+		err := app.container.Invoke(
+			func(log fxlog.Logger) error {
+				app.log = log
+				app.err = multierr.Combine(app.err, connectLogger(app.log))
+				if app.err != nil {
+					return app.err
+				}
+
+				return nil
```

Comment:
We shouldn't manipulate app.err inside the Invoke. If we got to the Invoke, we know app.err is blank, so we can just return the error.

```
err := Invoke(func(log fxlog.Logger) error {
  app.log = log
  return connectLogger(app.log)
})

if err != nil {
  app.err = err
  // ...
}
```

If there are any errors in invoking (because the logger could not be constructed), or because `connectLogger` failed, that'll be surfaced in `err`.